### PR TITLE
Add dependency graph generation for cycles

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
A new target `graph-deps` will generate a graph in `$O/deps.svg` showing all circular dependencies for the project. Using different options other kind of dependency graphs can be generated.

**Documentation is still missing but you can already try this**.

### CODE REVIEW

File: graph-deps [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        for x in nmark.iterkeys():
+            marked[x] = True
+    # remove all unmarked nodes
+    nodes = {}
+    for x in marked.iterkeys():
+        nodes[x.name] = x
+        x.adj = [i for i in x.adj if (i in marked)]
+
+# tarjan algorithm to find cycles
+# code borrowed from http://www.logarithmic.net/pfh/blog/01208083168
```
Comment (leandro-lucarella-sociomantic) [inline graph-deps:166]: Oh, damn, we have more licensing issues right here :worried:

**Comment by leandro-lucarella-sociomantic** [comment]
Blocking until I can contact the author of that code.

**Comment by leandro-lucarella-sociomantic** [comment]
E-mail sent.

**Comment by leandro-lucarella-sociomantic** [comment]
Paul Harrison replied:

> The code is derived from http://www.logarithmic.net/pfh/blog/01208083168 ?
> 
> No problem with the license. I want the code in the link above to be in the public domain, you can apply whatever license you like.

I looked at the code and the one in that post and they look quite different anyway, but license-wise we have definitively green light! Now I just need to update with proper documentation.

File: graph-deps (hunk 1 above)
Comment (leandro-lucarella-sociomantic) [reply graph-deps:166]: All good with this now!

**Comment by leandro-lucarella-sociomantic** [comment]
Updated.
- Added documentation
- Added entry in the release notes

This is ready for the final review/merge.

File: README.rst [hunk 2]
```diff
@@ -272,7 +272,7 @@ work ;). If you can't do that (because you generated a source file for example),
 you can use the special variable ``clean`` too (``clean += src/trash.d
 src/garbage.d`` for example).
 
-The ``doc`` itarget will, by default, call `harbored-mod
+The ``doc`` target will, by default, call `harbored-mod
```
Comment (nemanja-boric-sociomantic) [inline README.rst:276]: when vim leave the trail :-D

File: graph-deps [hunk 3]
```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python
+#
+# This script takes a dmd dependency file, and outputs a graph in graphviz
+# format also, detect cycles
```
Comment (nemanja-boric-sociomantic) [inline graph-deps]: `also, detect cycles` sounds weird, maybe:

```
This script takes a dmd dependency file, optionally detect cycles, and outputs a graph in graphviz format.
```

or something?

File: graph-deps [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+#
+# Based on: https://github.com/dellort/lumbricus/blob/d2/src/scripts/deps.py
+# Which used code from: http://www.logarithmic.net/pfh/blog/01208083168
```
Comment (nemanja-boric-sociomantic) [inline graph-deps:7]: Capital `W`

File: graph-deps [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+parser = OptionParser(usage="%prog [options] depfile graph.out",
+    description="Create a graphviz dependency graph from dmd module dependency file.",
+    add_help_option=False)
+parser.add_option("-h", "--help",
+    action="callback", callback=print_help,
+    help="show this help message and exit")
+parser.add_option("-c", "--cycles-only",
+    action="store_true", dest="cycles_only", default=False,
+    help="show only nodes, that are parts of a cycle")
```
Comment (nemanja-boric-sociomantic) [inline graph-deps]: extra `,`

File: graph-deps [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+parser.add_option("-C", "--cycle-edges-only",
+    action="store_true", dest="cycle_edges_only", default=False,
+    help="show only edges, that are part of a cycle")
```
Comment (nemanja-boric-sociomantic) [inline graph-deps]: extra `,`

File: graph-deps [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+parser.add_option("-e", "--exclude",
+    action="append", dest="ignore", default=[],
+    help="ignore module (or if it ends with a '.', ignore all packages starting with this)")
```
Comment (nemanja-boric-sociomantic) [inline graph-deps]: maybe? `(or ignore all packages starting with this value, if it ends with a`.`)`

File: graph-deps [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    help="show dependencies between packages")
+parser.add_option("--no-std-ignore",
+    action="store_true", dest="no_std_ignore", default=False,
+    help=("don't ignore standard modules (%s)" % std_ignore))
+parser.add_option("--imports",
+    action="append", dest="imports", default=[],
+    help="include only modules which import this module (transitively)")
+parser.add_option("--imports-depth",
+    action="store", type="int", dest="import_depth", default=0,
+    help="how many indirections max for --imports (0 means no restriction)")
```
Comment (nemanja-boric-sociomantic) [inline graph-deps]:
```
Maximum number of indirections for --imports (0 means no limit)
```

**Review by nemanja-boric-sociomantic — COMMENTED** [review]
On shallow look, looks good! I've put more attention reviewed `Makd` integration, and as well on `--help`. Would you care to fix these minor things?

File: graph-deps (hunk 3 above)
Comment (leandro-lucarella-sociomantic) [reply graph-deps]: Yeah, I think I left the original description untouched, but I'll change it.

File: graph-deps (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [reply graph-deps:7]: It was on purpose, they are like independent items. If you think is confusing I can change it though, but I will only after getting a confirmation.

Comment (nemanja-boric-sociomantic) [reply graph-deps:7]: No, that's fine.

File: graph-deps (hunk 5 above)
Comment (leandro-lucarella-sociomantic) [reply graph-deps]: It looks like the original author likes commas :)

Comment (leandro-lucarella-sociomantic) [reply graph-deps]: I perceive some Germanness in those extra commas :)

Comment (nemanja-boric-sociomantic) [reply graph-deps]: Good it's Germanness, because that preserves articles :-)

File: graph-deps (hunk 7 above)
Comment (leandro-lucarella-sociomantic) [reply graph-deps]: `ignore a module or an entire package (if it ends with a '.')`?

Comment (nemanja-boric-sociomantic) [reply graph-deps]: yeah, much better!

**Comment by leandro-lucarella-sociomantic** [comment]
Updated with the fixes/suggestions.

Note that I've done them in another commit to keep separated the original script and our modifications. I'll create another PR with many more small fixes or improvements to the script.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by nemanja-boric-sociomantic** [comment]
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was 693aaaff65bd442c5ed707f3baad99d514067ebb, new (rebased) HEAD is 58dae9efc47762f85b52124becbb8e9dc1d1e16a

**Comment by gavin-norman-sociomantic** [comment]
Awesome! That went pretty quickly after all.

**Pull request closed without merging**